### PR TITLE
Deluxe tags.

### PR DESCRIPTION
**Pull request by swankjesse**
This expands our single Object.class into a Map<Class<T>, T>.

Closes: https://github.com/square/okhttp/issues/4087

### CODE REVIEW

File: okhttp/src/main/java/okhttp3/Request.java [hunk 1]
```diff
@@ -239,12 +262,29 @@ public Builder method(String method, @Nullable RequestBody body) {
...
+     * request using {@link Request#tag}. Use null to remove any existing tag assigned for {@code
+     * type}.
+     *
+     * <p>Use this API to attach timing, debugging, or other application data to a request so that
+     * you may read it in interceptors, event listeners, or callbacks.
      */
-    public Builder tag(Object tag) {
-      this.tag = tag;
+    public <T> Builder tag(Class<? extends T> type, @Nullable T tag) {
+      if (type == null) throw new IllegalArgumentException("type == null: " + type);
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Request.java]: NPE. No point in passing the value to exception

File: okhttp/src/main/java/okhttp3/Request.java [hunk 2]
```diff
@@ -66,8 +69,24 @@ public Headers headers() {
...
   public Object tag() {
-    return tag;
+    return tag(Object.class);
+  }
+
+  /**
+   * Returns the tag attached with {@code type} as a key, or null if no tag is attached with that
+   * key.
+   */
+  public <T> T tag(Class<? extends T> type) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Request.java]: Nullable

File: okhttp-tests/src/test/java/okhttp3/RequestTest.java [hunk 3]
```diff
@@ -242,6 +243,103 @@ private void assertForbiddenHeader(String s) {
...
+  }
+
+  @Test public void defaultTag() {
+    UUID tag = UUID.randomUUID();
+    Request request = new Request.Builder()
+        .url("https://square.com")
+        .tag(tag)
+        .build();
+    assertEquals(tag, request.tag());
+    assertEquals(tag, request.tag(Object.class));
```
Comment (JakeWharton) [inline okhttp-tests/src/test/java/okhttp3/RequestTest.java]: If I'm being super nit picky these should all be assertSame

File: okhttp/src/main/java/okhttp3/Request.java [hunk 4]
```diff
@@ -66,8 +69,24 @@ public Headers headers() {
...
 
+  /**
+   * Returns the tag attached with {@code Object.class} as a key, or null if no tag is attached with
+   * that key.
+   *
+   * <p>Prior to OkHttp 3.11, this method never returned null if no tag was attached. Instead it
+   * returned either this request, or the request upon which this request was derived with {@link
+   * #newBuilder()}.
+   */
   public Object tag() {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Request.java]: Nullable

File: okhttp/src/main/java/okhttp3/Request.java [hunk 5]
```diff
@@ -239,12 +262,29 @@ public Builder method(String method, @Nullable RequestBody body) {
...
+     * Attaches {@code tag} to the request using {@code type} as a key. Tags can be read from a
+     * request using {@link Request#tag}. Use null to remove any existing tag assigned for {@code
+     * type}.
+     *
+     * <p>Use this API to attach timing, debugging, or other application data to a request so that
+     * you may read it in interceptors, event listeners, or callbacks.
      */
-    public Builder tag(Object tag) {
-      this.tag = tag;
+    public <T> Builder tag(Class<? extends T> type, @Nullable T tag) {
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Request.java]: My only complaint about this is that I can't easily store generic types or multiple of the same type. You can create wrapper or subclasses for statically known tagging, but you can dynamically add multiple. I don't know if that's something we care about probably being a `Map<String, Object>` where typing is the user problem.

File: okhttp/src/main/java/okhttp3/Request.java [hunk 6]
```diff
@@ -239,12 +262,29 @@ public Builder method(String method, @Nullable RequestBody body) {
...
+      if (type == null) throw new IllegalArgumentException("type == null: " + type);
+
+      if (tag == null) {
+        tags.remove(type);
```
Comment (JakeWharton) [inline okhttp/src/main/java/okhttp3/Request.java]: Is this safe on empty map?

Comment (PromanSEW) [reply okhttp/src/main/java/okhttp3/Request.java]: No, `UnsupportedOperationException` is thrown for `Collections.emptyMap()`.
This method returns unmodifiable empty map.
This operation is safe only for empty `LinkedHashMap`

File: okhttp/src/main/java/okhttp3/Request.java (hunk 5 above)
Comment (yschimke) [reply okhttp/src/main/java/okhttp3/Request.java]: Main use cases I see

- oauth tokens
- tracing context flags
- logging tag set (zerorated, video, userinteractive, experimentlabels)

I think for any of these I don't expect that there will be hundreds of call sites in an app setting these on the request, so client dealing with String -> Object is ok.

File: okhttp/src/main/java/okhttp3/Request.java (hunk 1 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Request.java]: Fixed. Too much Kotlin lately!

File: okhttp/src/main/java/okhttp3/Request.java (hunk 2 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Request.java]: Fixed x 2.

File: okhttp-tests/src/test/java/okhttp3/RequestTest.java (hunk 3 above)
Comment (swankjesse) [reply okhttp-tests/src/test/java/okhttp3/RequestTest.java]: Good call. Fixed all.

File: okhttp/src/main/java/okhttp3/Request.java (hunk 4 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Request.java]: Fixed.

File: okhttp/src/main/java/okhttp3/Request.java (hunk 6 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Request.java]: It’s safe. I added this test.
```
  @Test public void removeAbsentTag() {
    Request request = new Request.Builder()
        .url("https://square.com")
        .tag(null)
        .build();
    assertNull(request.tag());
  }
```

File: okhttp/src/main/java/okhttp3/Request.java (hunk 5 above)
Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Request.java]: I’m imagining similar enough use cases, but I’m assuming boxing strings isn’t much of a hassle. I also think we could leverage existing types in tags:

```
Request request = ...
retrofit2.Call<?> retrofitCall = request.tag(retrofit2.Call.class);
zipkin.Span span = request.tag(zipkin.Span.class);
...
```

**Comment by adriancole** [comment]
main comment is I like the feature name.

File: okhttp/src/main/java/okhttp3/Request.java (hunk 5 above)
Comment (adriancole) [reply okhttp/src/main/java/okhttp3/Request.java]: I like that we can do this.. quite a lot.

File: okhttp/src/main/java/okhttp3/Request.java [hunk 7]
```diff
@@ -239,12 +262,29 @@ public Builder method(String method, @Nullable RequestBody body) {
...
-    public Builder tag(Object tag) {
-      this.tag = tag;
+    public <T> Builder tag(Class<? extends T> type, @Nullable T tag) {
+      if (type == null) throw new NullPointerException("type == null");
+
+      if (tag == null) {
+        tags.remove(type);
+      } else {
+        if (tags.isEmpty()) tags = new LinkedHashMap<>();
+        tags.put(type, tag);
```
Comment (stevecstian) [inline okhttp/src/main/java/okhttp3/Request.java]: Should we have a dynamic cast here to achieve runtime type safety?

Comment (swankjesse) [reply okhttp/src/main/java/okhttp3/Request.java]: Good idea. Done.

File: okhttp/src/main/java/okhttp3/Request.java (hunk 5 above)
Comment (stevecstian) [reply okhttp/src/main/java/okhttp3/Request.java]: A tiny wrapper should help for generic type:
```java
interface MyListSupplier extends Supplier<List<String>> {}
...
List<String> myList = request.tag(ListSupplier.class).get();
...
```

Comment (JakeWharton) [reply okhttp/src/main/java/okhttp3/Request.java]: Assuming you have the flexibility to subtype and not wrap. And assuming your entries are statically known and not dynamic.

Comment (stevecstian) [reply okhttp/src/main/java/okhttp3/Request.java]: Yup

**Pull request merged**